torch/csrc/jit/passes/onnx/shape_inference.cpp

```diff
@@ -0,0 +1,168 @@
+#include <torch/csrc/jit/passes/onnx/shape_inference.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/onnx/helper.h>
+#include <torch/csrc/jit/serialization/export.h>
+#include <torch/csrc/jit/serialization/onnx.h>
+
+#include <onnx/shape_inference/implementation.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+namespace onnx_torch = ::torch::onnx;
+namespace onnx = ::ONNX_NAMESPACE;
+
+void UpdateValueByValueInfo(Value* v, const onnx::ValueInfoProto& p_info) {
+  if (!p_info.has_type()) {
+    return;
+  }
+
+  auto p_type = p_info.type();
+  if (!p_type.has_tensor_type()) {
+    // TODO: Support sequence type.
+    return;
+  }
+
+  auto p_tensor_type = p_type.tensor_type();
+
+  c10::optional<at::ScalarType> scalar_type;
+  if (p_tensor_type.has_elem_type()) {
+    scalar_type = ONNXTypeToATenType(p_tensor_type.elem_type());
+  }
+
+  auto v_type = TensorType::create(
+      scalar_type,
+      at::kCPU,
+      c10::SymbolicShape(),
+      c10::VaryingShape<c10::Stride>{},
+      {});
+  if (p_tensor_type.has_shape()) {
+    std::vector<int64_t> sizes;
+    auto p_shape = p_tensor_type.shape();
+
+    for (int i = 0; i < p_shape.dim_size(); ++i) {
+      auto& dim = p_shape.dim(i);
+      if (dim.has_dim_value()) {
+        sizes.push_back(dim.dim_value());
+      } else {
+        // TOOD: handle dim_param?
+        return;
+      }
+    }
+    v_type = TensorType::create(scalar_type, at::kCPU, sizes.size(), {});
+    v_type = v_type->withSizes(sizes);
+  }
+
+  GRAPH_UPDATE("Prev Value: ", *(v->node()));
+  v->setType(v_type);
+  GRAPH_UPDATE("Update Value: ", *(v->node()));
+}
+
+bool IsSupportedNode(const Node* n) {
+  auto node_kind = n->kind();
+
+  if (!node_kind.is_onnx()) {
+    // node kind is not ONNX, skipped.
+    return false;
+  }
+
+  if (node_kind == ::c10::onnx::SequenceAt ||
+      node_kind == ::c10::onnx::SplitToSequence ||
+      node_kind == ::c10::onnx::SequenceConstruct ||
+      node_kind == ::c10::onnx::SequenceEmpty ||
+      node_kind == ::c10::onnx::SequenceInsert ||
+      node_kind == ::c10::onnx::ConcatFromSequence ||
+      node_kind == ::c10::onnx::SequenceErase) {
+    // TODO: ONNX unable to do shape inference for these ops.
+    return false;
+  }
+
+  if (node_kind == ::c10::onnx::ConstantOfShape) {
+    // && n->input()->node()->kind() == ::c10::prim::ListConstruct
+    // TODO: ONNX shape inference segfault.
+    return false;
+  }
+
+  if (node_kind == ::c10::onnx::Loop || node_kind == ::c10::onnx::If) {
+    // TODO: Support Loop & If shape inference by propagating input shape to
+    // block input.
+    return false;
+  }
+
+  return true;
+}
+
+} // namespace
+
+void ONNXShapeInference(Node* n, int opset_version) {
+  if (!IsSupportedNode(n)) {
+    return;
+  }
+
+  auto node_kind = n->kind();
+
+  GRAPH_UPDATE("Shape inference for node kind: ", node_kind.toDisplayString());
+  auto g = n->owningGraph();
+
```

Comment:
Can we add a comment here to describe what is being done here. So something like "Creating a separate graph for just one node".
In fact, it will be good to have a comment describing the different steps being done here at the top and have more comments at each step interspersed in the code (like here).